### DISCUSSION TITLE
ports/Interface defined but not used in entire project

### DISCUSSION THREAD

**Discussion by jolosastro** [LordMoMA/Hexagonal-Architecture#6] [Q&A] [answered]
There are ports for services (example ports.UserService, ports.MessengerService), but I don't see it used in the project. Why are they defined but not used?

**Comment by cycloss** [comment, marked as answer]
I think they ought to be being used. For example in `internal/adapters/handler/user_handler.go`, the `UserHandler` is referencing the concrete implementation of `UserService` in the `services` package, but I think it should instead reference the `UserService` in the `ports` package.

The advantage of this will be increased decoupling, and will allow the `UserService` to be switched out for a different implementation. For example, this might be useful for writing tests for testing the decoding and validating http requests where a mock service can be injected. There is also a niche use case which I encountered recently where there are two types of user who's type dictates which of one or two third party services their data resides on. In this case it is useful for the handler to only have an understanding of the interface defined in the `ports` package, and for a factory method to return a specific implementation to the handler depending on the user type.
